compiler/rustc_attr/src/builtin.rs

```diff
@@ -839,22 +860,30 @@ pub fn find_deprecation(
             }
         }
 
-        if is_rustc {
-            if since.is_none() {
-                sess.emit_err(session_diagnostics::MissingSince { span: attr.span });
-                continue;
+        let since = if let Some(since) = since {
+            if since.as_str() == "TBD" {
+                Some(DeprecatedSince::Future)
+            } else if !is_rustc {
+                Some(DeprecatedSince::Symbol(since))
+            } else if let Some(version) = parse_version(since) {
+                Some(DeprecatedSince::RustcVersion(version))
+            } else {
+                sess.emit_err(session_diagnostics::InvalidSince { span: attr.span });
+                Some(DeprecatedSince::Symbol(since))
             }
+        } else if is_rustc {
+            sess.emit_err(session_diagnostics::MissingSince { span: attr.span });
+            continue;
```

Comment:
Should we stop discarding the attribute on missing since?

Reply:
Yes, good catch. Fixed.

I chose to use `DeprecatedSince::Err` here instead of `DeprecatedSince::Unspecified`. I used `Unspecified` for the non-`feature(staged_api)` case where deprecation versions are optional, and `DeprecatedSince::Err` for the `feature(staged_api)` case where deprecation versions are mandatory and must be parsed as a valid version; whether missing or invalid, you get `DeprecatedSince::Err`, and it tells downstream code that it can rely on an error already having been emitted. Let me know what you think.